### ISSUE TITLE
Allow vendor service file that are not enabled by default

### ISSUE THREAD

**Issue by DeX77** [troglobit/finit#71]
Hi,

referring to the discussion in #69 a way to provide some service definition without automatically enabling it would be nice.

Traditional inits use a file based approach for that: if a symlink to the actual config is found in a runlevel/target specific directory the service will be enabled.

This is "UNIX way" and its easy for a Distro to provide those files, and have a script to either enable the service on install or use a *ctl tool to do it.

The other idea mentioned in #69 (to use a command to create the file) is also possible but not _that_ easy.

Also it might be an idea to have a "override" feature so files in `/etc/finit.d` will override those in `/usr/lib/finit.d` which would be difficult with the command.

**Comment by troglobit** [comment]
Hi!

Yes, this is what I'm considering. Just discussed this with a colleague of mine and what we came up with is something like this:

1. Define a *recommended* directory structure for distributions
2. Document better the possibility to change the location of `/etc/finit.d` to whatever a distribution likes
3. Update man pages
4. Provide examples

Currently I'm leaning towards this layout:

    /etc/finit/
      |-- available.d/*.conf
      `-- enabled.d/dropbear.conf --> ../available.d/dropbear.conf
    /etc/finit.conf

A distribution could then use a script that symlinks (as in the above example), removes symlinks, like Apache `a2ensite`. It could also copy, create from XML/JSON/db, checkout from a version control system, or whatever feels suitable to their domain.

The default location which finit reads, after `/etc/finit.conf`, is possible to change with `configure --with-rcsd=/path/to/dot.d` and `/etc/finit.conf` can be reduced to a few tasks/services needed for bootstrapping the system, but the name of the file itself can also be changed using `configure --with-config=CONFIG`.

**Comment by DeX77** [comment]
Yes, I agree with that aproach and I think its quite intuitive for the usual Unix/Linux user.  

Perhaps, if I find some time I'll try to patch up something, But no promises. :wink:

**Comment by troglobit** [comment]
Just FYI, my primary use-case for Finit is what is encoded in `configure` and `*.h` files.  It is unlikely that I'd be in favor of changing these defaults :)

If you want to contribute anything for this particular issue I'd recommend writing a `DISTRIBUTION.md`, or similar, with instructions for packagers.  And maybe we could update the Debian and Alpine Linux samples.

Cheers

**Comment by abucodonosor** [comment]
@troglobit 

sure defaults ca be to whatever you wish :)

You define whatever it will be default and we can add configure options and some docs for that.

so we can do 

--with-some-dir=/foo ( for a distro ) but *the default* will be whatever you want

is similar with the discuss from PR #70.. Everything is a configure option for checking , setting up
prog PATHs can be defined like this 

--with-something=auto,PATH  default=/your_default_from_source_code_like_is_now

meaning the PATH or whatever only changes when someone want it too change 
and without the option set it will be alyway 'your_default'

Also I think with a apache like directory structure we can have some sort 
do_enable_service() / do_disable_service() in initctl.c ? Since is a matter of
check service  file exists , check is valid , if both yes symlink it and reload ?

**Comment by troglobit** [comment]
I've spent some time thinking about this, and also discussed with a colleague of mine.  I'm no longer in favor of adding support to `initctl` for enable/disable a service, that should be a separate tool/script, like Apache `a2ensite`/`a2dissite`.  

Introducing enable/disable on top of the existing stop/start would be highly confusing.  Also, enable/disable will be very distribution specific, so better keep it separate.  However, I *am* in favor of having such a script in-tree for convenience.

**Comment by troglobit** [comment]
Although ... systemd has this enable/disable madness included in their `systemctl`, along with start/stop ... dang it, this is why we can't have nice things, like usability and predictability!

Gah! :stuck_out_tongue_winking_eye: 

OK, if someone provides a neat pull-request and *maybe even changes the defaults* ... as long as it's symmetric, well indented code, who am I to argue?  At least I'm open to discussing around a piece of code ... it's always easier if there's code :smiley:

**Comment by troglobit** [comment]
Things have been quiet in this thread ever since I displayed my ambivalence in this matter. Time is a scarce resource and should not be spent implementing something with the risk of a maintainer then nixing it ... my apologies.

Nevertheless, I've been thinking about this on and off for the last couple of months. Trying to get a feel for what is the right namespace, how it can be implemented without breaking the existing workflow, etc.

What I came up with is a simplified variant of what I proposed early on:

    /etc/finit.conf
    /etc/finit.d/
      |-- available/
      |     |-- dropbear.conf
      |     |-- mini-snmpd.conf
      |     `-- uftpd.conf
      |-- uftpd.conf --> available/uftpd.conf
      `-- dropbear.conf --> available/dropbear.conf

In 3428cd4 I've implemented this as three `initctl` commands:
- `initctl list`: List all services, available and enabled
- `initctl enable <CONF>`: Enable a service by adding a symlink from available/ to the parent dir
- `initctl disable <CONF>`: Disable a service by removing its symlink

### RFC

I'd very much like to get feedback on this. Have I missed something obvious? Does anyone have any questions regarding how it can be used or tweaked?

### Example

The below example run is from [TroglOS](https://github.com/troglobit/troglos)

```sh
root@chaos:~ $ initctl list
/etc/finit.d/available ========================================================
http.conf       inetd.conf      lldpd.conf      ntpd.conf       rdate.conf      
snmpd.conf      sshd.conf       telnetd.conf    uftpd.conf      watchdogd.conf  

/etc/finit.d ==================================================================
lldpd.conf  

root@chaos:~ $ initctl enable snmpd
root@chaos:~ $ initctl list
/etc/finit.d/available ========================================================
http.conf       inetd.conf      lldpd.conf      ntpd.conf       rdate.conf      
snmpd.conf      sshd.conf       telnetd.conf    uftpd.conf      watchdogd.conf  

/etc/finit.d ==================================================================
lldpd.conf  snmpd.conf  

root@chaos:~ $ initctl enable uftpd
root@chaos:~ $ initctl list
/etc/finit.d/available ========================================================
http.conf       inetd.conf      lldpd.conf      ntpd.conf       rdate.conf      
snmpd.conf      sshd.conf       telnetd.conf    uftpd.conf      watchdogd.conf  

/etc/finit.d ==================================================================
lldpd.conf  snmpd.conf  uftpd.conf  

root@chaos:~ $ initctl enable ntpd.conf
root@chaos:~ $ initctl list
/etc/finit.d/available ========================================================
http.conf       inetd.conf      lldpd.conf      ntpd.conf       rdate.conf      
snmpd.conf      sshd.conf       telnetd.conf    uftpd.conf      watchdogd.conf  

/etc/finit.d ==================================================================
lldpd.conf  ntpd.conf   snmpd.conf  uftpd.conf  

root@chaos:~ $ initctl reload
root@chaos:~ $ initctl status
#      Status   PID     Runlevels   Service               Description
===============================================================================
1      running  90      [S12345]    syslogd               System log daemon    
2      running  91      [S12345]    klogd                 Kernel log daemon    
3      running  120     [2345]      lldpd                 LLDP daemon (IEEE 802
4      running  138     [2345]      ntpd                  NTP client/server    
5      running  137     [2345]      mini_snmpd            Mini snmpd           
6      running  inetd   [2345]      uredir                UDP port redirector  
7      running  inetd   [2345]      in.ftpd               FTP daemon           
8      running  inetd   [2345]      in.tftpd              TFTP daemon          

root@chaos:~ $ initctl disable ntpd
root@chaos:~ $ initctl list
/etc/finit.d/available ========================================================
http.conf       inetd.conf      lldpd.conf      ntpd.conf       rdate.conf      
snmpd.conf      sshd.conf       telnetd.conf    uftpd.conf      watchdogd.conf  

/etc/finit.d ==================================================================
lldpd.conf  snmpd.conf  uftpd.conf  
```

**Comment by troglobit** [comment]
Merged to master branch for upcoming v3.0 release.
